ambry-messageformat/src/main/java/com.github.ambry.messageformat/BlobStoreHardDelete.java

```diff
@@ -87,6 +90,43 @@ public MessageInfo getMessageInfo(Read read, long offset, StoreKeyFactory storeK
             return new MessageInfo(key, header.capacity() + key.sizeInBytes() + headerFormat.getMessageSize(), true,
                 deleteRecord.getAccountId(), deleteRecord.getContainerId(), deleteRecord.getDeletionTimeInMs());
           }
+        }
+        case MessageFormatRecord.Message_Header_Version_V2: {
+          header = ByteBuffer.allocate(MessageFormatRecord.MessageHeader_Format_V2.getHeaderSize());
+          header.putShort(version);
+          read.readInto(header, offset);
+          header.flip();
+          MessageFormatRecord.MessageHeader_Format_V2 headerFormatV2 =
+              new MessageFormatRecord.MessageHeader_Format_V2(header);
+          headerFormatV2.verifyHeader();
+          boolean hasEncryptionKey = headerFormatV2.getBlobEncryptionKeyRecordRelativeOffset()
+              != MessageFormatRecord.Message_Header_Invalid_Relative_Offset;
+          offset += header.capacity() - headerVersion.capacity();
+          endOffset = headerFormatV2.getBlobPropertiesRecordRelativeOffset()
+              != MessageFormatRecord.Message_Header_Invalid_Relative_Offset ? offset + (hasEncryptionKey
+              ? headerFormatV2.getBlobEncryptionKeyRecordRelativeOffset()
+              : headerFormatV2.getBlobPropertiesRecordRelativeOffset()) + headerFormatV2.getMessageSize()
+              : offset + headerFormatV2.getDeleteRecordRelativeOffset() + headerFormatV2.getMessageSize();
+
+          stream = new ReadInputStream(read, offset, endOffset);
+          StoreKey key = storeKeyFactory.getStoreKey(new DataInputStream(stream));
+          if (hasEncryptionKey) {
+            MessageFormatRecord.deserializeBlobEncryptionKey(stream);
+          }
+
+          // read the appropriate type of message based on the relative offset that is set
+          if (headerFormatV2.getBlobPropertiesRecordRelativeOffset()
```

Comment:
I did that wherever I thought it could be done in a clean way. Sometimes it didn't make sense to cherry-pick a few common lines and add a method for it.